File: src/frameserver/remoting/default/remoting.c
/*
 * Arcan Remoting reference Frameserver
 * Copyright 2014-2017, Björn Ståhl
 * License: 3-Clause BSD, see COPYING file in arcan source repository.
 * Reference: http://arcan-fe.com
 * Depends: libvncserver (GPLv2)
 * Description: This is a 'quick and dirty' default implementation of a
 * remoting client, it's not very feature complete and only covers the RFB
 * protocol for now.
 */
#include "arcan_shmif.h"
#include "remoting.h"

int afsrv_remoting(struct arcan_shmif_cont* con, struct arg_arr* args)
{
	int rv = EXIT_FAILURE;
	if (!con)
		return rv;

	const char* protocol = NULL;
	arg_lookup(args, "protocol", 0, &protocol);

	if (!protocol || strcasecmp(protocol, "vnc") == 0){
#ifdef ENABLE_VNC
		rv = run_vnc(con, args);
#else
		arcan_shmif_last_words(con, "Arcan was not built with VNC support");
		fprintf(stderr, "VNC protocol disabled\n");
#endif
	}
	else if (strcasecmp(protocol, "a12") == 0){
		rv = run_a12(con, args);
	}
	else {
		char buf[64];
		snprintf(buf, sizeof(buf), "Unknown protocol (%s)", protocol);
		arcan_shmif_last_words(con, buf);
		fprintf(stderr, "%s", buf);
	}

	return rv;
}

File: src/frameserver/remoting/default/remoting.zig
// Zig port of remoting.c — remoting frameserver dispatcher
const std = @import("std");
const shmif = @import("shmif_types");
const libc = @import("posix");

/// Dispatch struct stitching together every C-namespace symbol the file
/// uses, routed to the right replacement module.
const c = struct {
    pub const arcan_shmif_cont = shmif.arcan_shmif_cont;
    pub const arg_arr = shmif.arg_arr;
    pub const arg_lookup = shmif.arg_lookup;
    pub const arcan_shmif_last_words = shmif.arcan_shmif_last_words;
    pub const fprintf = libc.fprintf;
    // Re-declare extern var (see encode.zig comment).
    pub extern "c" var stderr: *libc.FILE;
};

extern "c" fn run_a12(cont: ?*c.arcan_shmif_cont, args: ?*c.arg_arr) c_int;

export fn afsrv_remoting(con: ?*c.arcan_shmif_cont, args: ?*c.arg_arr) callconv(.c) c_int {
    if (con == null) return 1;

    var protocol: [*c]const u8 = null;
    _ = c.arg_lookup(args, "protocol", 0, &protocol);

    if (protocol == null) {
        c.arcan_shmif_last_words(con, "Arcan was not built with VNC support");
        _ = c.fprintf(c.stderr, "VNC protocol disabled\n");
        return 1;
    }

    const proto = std.mem.span(protocol.?);
    if (std.ascii.eqlIgnoreCase(proto, "vnc")) {
        c.arcan_shmif_last_words(con, "Arcan was not built with VNC support");
        _ = c.fprintf(c.stderr, "VNC protocol disabled\n");
        return 1;
    } else if (std.ascii.eqlIgnoreCase(proto, "a12")) {
        return run_a12(con, args);
    } else {
        var buf: [64]u8 = undefined;
        const msg = std.fmt.bufPrintZ(&buf, "Unknown protocol ({s})", .{proto}) catch "Unknown protocol";
        c.arcan_shmif_last_words(con, msg.ptr);
        _ = c.fprintf(c.stderr, "%s", msg.ptr);
        return 1;
    }
}
